### ISSUE TITLE
in_app_purchase keep getting items to restore

### ISSUE THREAD

**Issue by vytautas-pranskunas-** [flutter/flutter#85529]
Hi,

I am getting items to ```restore``` and verifying them on server. After verification I am calling ```completePurchase``` but after app refresh still keep getting same items on ```restore```. This ghappens only on iOS.

Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 2.2.2, on macOS 11.4 20F71 darwin-x64, locale en-CH)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
[✓] Xcode - develop for iOS and macOS
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.2)
[✓] VS Code (version 1.57.1)
[✓] Connected device (3 available)

• No issues found!

**Comment by vytautas-pranskunas-** [comment]
I think most developers would appreciate if these issue is prioritizes because basically nor Android nor iOS flow does not work

**Comment by TahaTesser** [comment]
Hi @vytautas-pranskunas- 
Thanks for the patience, we were having technical issues when reproducing the issue, I can finally test the issue 

> I am getting items to restore and verifying them on server. After verification I am calling completePurchase but after app refresh still keep getting same items on restore with pendingCompletePurchase = true. This happens only on iOS.


After making the changes you suggested, I wanted to try to how many times restore is being called, and it gets many times then stops on iOS
```
[        ] Restarted application in 693ms.
[+5123 ms] flutter: purchaseUpdated listen
[   +1 ms] flutter: restored called
[        ] flutter: restored called
[        ] flutter: restored called
[        ] flutter: restored called
[        ] flutter: restored called
[        ] flutter: restored called
[        ] flutter: restored called
[        ] flutter: restored called
[        ] flutter: restored called
[        ] flutter: restored called
[        ] flutter: restored called
[        ] flutter: restored called
[        ] flutter: restored called
[        ] flutter: restored called
[        ] flutter: restored called
[        ] flutter: restored called
[        ] flutter: restored called

```
When completing the purchase, refreshed the app button is green, when i try to purchase again I get the following
```
[+1508 ms] [VERBOSE-2:ui_dart_state.cc(199)] Unhandled Exception: PlatformException(storekit_duplicate_product_object,
There is a pending transaction for the same product identifier. Please either wait for it to be finished or finish it
manually using `completePurchase` to avoid edge cases., {applicationUsername: null, requestData: null, quantity: 1,
productIdentifier: com.tahatesser.triage.test_silver, simulatesAskToBuyInSandbox: false}, null)
[   +1 ms] #0      StandardMethodCodec.decodeEnvelope (package:flutter/src/services/message_codecs.dart:597:7)
[        ] #1      MethodChannel._invokeMethod (package:flutter/src/services/platform_channel.dart:158:18)
[        ] <asynchronous suspension>
[        ] #2      SKPaymentQueueWrapper.addPayment
(package:in_app_purchase_ios/src/store_kit_wrappers/sk_payment_queue_wrapper.dart:91:5)
[        ] <asynchronous suspension>
[        ] #3      InAppPurchaseIosPlatform.buyNonConsumable
(package:in_app_purchase_ios/src/in_app_purchase_ios_platform.dart:63:5)
[        ] <asynchronous suspension>
```

|  `stable`  | `master` |
| :---: |:---: |
|  ❌  |   ❌  | 

Check `flutter doctor -v` outputs for each channel below

<details> 
<summary>flutter doctor -v</summary> 
	
```console
[✓] Flutter (Channel stable, 2.2.2, on macOS 11.4 20F71 darwin-x64, locale
    en-US)
    • Flutter version 2.2.2 at /Users/tahatesser/Code/flutter_stable
    • Framework revision d79295af24 (3 weeks ago), 2021-06-11 08:56:01 -0700
    • Engine revision 91c9fc8fe0
    • Dart version 2.13.3

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Volumes/Extreme/SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Volumes/Extreme/SDK
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.5, Build version 12E262
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] VS Code (version 1.57.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.23.0

[✓] Connected device (3 available)
    • Taha’s iPhone (mobile) • 00008020-001059882212002E • ios            • iOS
      14.6
    • macOS (desktop)        • macos                     • darwin-x64     •
      macOS 11.4 20F71 darwin-x64
    • Chrome (web)           • chrome                    • web-javascript •
      Google Chrome 91.0.4472.114

• No issues found!
```

```console
[✓] Flutter (Channel master, 2.3.0-17.0.pre.578, on macOS 11.4 20F71 darwin-x64,
    locale en-US)
    • Flutter version 2.3.0-17.0.pre.578 at
      /Users/tahatesser/Code/flutter_master
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 01e10e793e (16 hours ago), 2021-06-29 20:41:02 -0400
    • Engine revision 9d2377f0d8
    • Dart version 2.14.0 (build 2.14.0-256.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Android SDK at /Volumes/Extreme/SDK
    • Platform android-30, build-tools 30.0.3
    • ANDROID_HOME = /Volumes/Extreme/SDK
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.5, Build version 12E262
    • CocoaPods version 1.10.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.8+10-b944.6916264)

[✓] VS Code (version 1.57.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.23.0

[✓] Connected device (3 available)
    • Taha’s iPhone (mobile) • 00008020-001059882212002E • ios            • iOS
      14.6 18F72
    • macOS (desktop)        • macos                     • darwin-x64     •
      macOS 11.4 20F71 darwin-x64
    • Chrome (web)           • chrome                    • web-javascript •
      Google Chrome 91.0.4472.114

• No issues found!
```
		
</details>

✅ : No Issue ❌: Issue reproduced

**Comment by vytautas-pranskunas-** [comment]
@TahaTesser Thanks for taking this. Plugin also does not work on Android: https://github.com/flutter/flutter/issues/85448
Please take this also up because (not sure about others) but i cannot complete in app purchases :(

Can you gues on time frame when can we expect both issues are fixed?

**Comment by vytautas-pranskunas-** [comment]
I wonder why this issue received P4 (as I understand P4 is lowest priority bug?). This plugin cannot be used with this bug because purchases cannot be made!

**Comment by msegmen** [comment]
I am experiencing this as well. Any updates on this?

**Comment by sivaprasadnk** [comment]
Me too stuck in this issue.

**Comment by vytautas-pranskunas-** [comment]
Hi @TahaTesser ,
can you comment why it take so long to fix it? Without this fix this library is not usable and there is no good, reliable alternative!
I wonder why so low priority has been given...

**Comment by ajaygautam** [comment]
Stuck with the same issue.
Please fix this ASAP!

**Comment by nosmirck** [comment]
I'm experiencing the exact same issue. it's impossible to restore purchases like this nor buy a new subscription because it is showing as still pending! the plugin is totally useless like this.

**Comment by kainjinez** [comment]
I temporarily fixed it by passing a variable, that prevents handing multiple restore events in a single list. Hope developer can fix it and also other major bugs. This library is coming to be unstable & useless.

**Comment by etienn01** [comment]
> I temporarily fixed it by passing a variable, that prevents handing multiple restore events in a single list. Hope developer can fix it and also other major bugs. This library is coming to be unstable & useless.

Hi @kainjinez, I'm facing the same issue, could you give more details on how you implemented your temporary fix?
Thanks!

**Comment by akshaytaru2007** [comment]
I m also facing the same issue. Would really appreciate if you can provide some solution.

**Comment by ajaygautam** [comment]
Checkout Revenue Cat - Its a much simpler implementation that works out of the box!
It took me a few hours to implement with RevenueCat after wasting 2 weeks with in_app_purchase.
Good luck.

**Comment by Kossi-Francois** [comment]
in your "purchase" (or "buy") method, in the case of a non-consumable, you have to finish old transaction on ios platform before calling iap.buyNonConsumable like [this](https://github.com/flutter/flutter/issues/79473#issuecomment-842245471) :

import `import 'package:in_app_purchase_ios/store_kit_wrappers.dart'; `

```
case storeKeyUpgrade:
      if (Platform.isIOS) {
        print("clearing past subscriptions");
        final paymentWrapper = SKPaymentQueueWrapper();
        final transactions = await paymentWrapper.transactions();
        await Future.wait(transactions
            .map((transaction) => paymentWrapper.finishTransaction(transaction)));
      }
        await iapConnection.buyNonConsumable(purchaseParam: purchaseParam);
```

and also in the "main.dart",  you should have a plateforme check like [this](https://github.com/flutter/flutter/issues/89950#issuecomment-921594203) as in the documentation: 

```
if (defaultTargetPlatform == TargetPlatform.android) {
    InAppPurchaseAndroidPlatformAddition.enablePendingPurchases();
}
runApp(MyApp());
```

I tested (with the last version of IAP  in_app_purchase: ^1.0.9) and it seems to work on Android as on IOS.

**Comment by sleith** [comment]
Hi @akshaytaru2007 
1. yes
2. yes as well to automatically load all the purchases he has
Also i want to know when the purchase has expired (cancelled), without having server. So i was hoping with restore purchase can see the current purchases which are still valid.
Let me know if there's a better method to achieve this. 
Thank you

**Comment by akshaytaru2007** [comment]
@sleith 

I am pretty sure you are not gonna like my answer but this is not at all possible without server. 

If you click 10times or restore purchases on iOS. You will get 10 products from purchase stream.

I had bang my head on this and finally I am using Google cloud function as API and firestore as No sql db. As temporary solution.

But I will surely move to my own server to have more control.

Also, you will only get expiry date from Server verification (either receipt or original transaction id).

Basically you have to call 2 API's

1. Verify receipt (while purchase is made, it will give you original transaction id also with all details and expiry date)
2. Verify subscription. Whenever app is launch you can verify subscription using original transaction id. 

I am happy to help if you need guide.

**Comment by sleith** [comment]
Hi @akshaytaru2007 
Thanks for your info. This question is out of this thread question, but would it possible to let me know the proper steps for the app to handle subscription, restore, and validate? 

I read `https://developer.apple.com/documentation/storekit/original_api_for_in-app_purchase/choosing_a_receipt_validation_technique` can be verified locally and server side. 
But not sure with local method, if user change other device, he needs to restore purchase first or how to verify this (the receipt is on other device). 
Or if need server side, which data to send to server from this SKPaymentTransactionWrapper object?

For Android case, is it safe to restore purchase many times? i tried it works just fine, but not sure if any side effects.
Thanks a lot for your help

**Comment by akshaytaru2007** [comment]
For iOS Local verification is possible with Native code. 

On server you have to send receipt (serverVerificationData) to verify. 

 Link to API:

Verify receipt:
https://developer.apple.com/documentation/appstorereceipts/verifyreceipt

For Android you can call restore API as much you want but you won't get expiry date :(

**Comment by Kossi-Francois** [comment]
@sleith,  @akshaytaru2007 is right, you have to use a server, I advise you to look at the official [example](https://codelabs.developers.google.com/codelabs/flutter-in-app-purchases#0) (it's a codelab) which uses Firebase, it's pretty well done.

**Comment by KlausJokisuo** [comment]
@TahaTesser Any news on this or some workaround?

**Comment by mdmota** [comment]
Can someone help me. On Android everything is ok.
I'm having trouble testing on iOS. the error happens in both the emulator and the device

I already have my sandbox user registered. When running the App in the emulator, when I call InAppPurchase.instance.restorePurchases() this Sig in window appears.
![aaa](https://user-images.githubusercontent.com/21266037/143719326-784ac299-79b1-4a00-9b5b-7c96c122207a.jpg)

I've already tried to enter the sandbox username and password but the console gives an error:
Instance of 'SKError'
![xxxx](https://user-images.githubusercontent.com/21266037/143719332-5c0fb25b-b903-49e4-94e0-169713ac881d.jpg)

in_app_purchase: ^1.0.9

**Comment by Kossi-Francois** [comment]
@KlausJokisuo
im using [this](https://github.com/flutter/flutter/issues/85529#issuecomment-937930673) workaround and it seems to work fin, you can try and let me know if it ok

**Comment by greymag** [comment]
Any update on this? The plugin is unusable for subscriptions with this bug.

**Comment by pro100svitlo** [comment]
Shame!

P4 ???? Really?

**Comment by pro100svitlo** [comment]
I find out a another workaround, which helped me with `PlatformException(storekit_duplicate_product_object)`.
I describe it in this [issue](https://github.com/flutter/flutter/issues/94011#issuecomment-1040110848)
Please try it and let me know if that works for you as well 😉

**Comment by tfoel** [comment]
I'm wondering how this code in the StoreKit wrapper makes any sense:
![image](https://user-images.githubusercontent.com/1913456/168239606-9d580f24-e2b7-49a5-b626-06b378c81d6e.png)

https://github.com/flutter/plugins/blob/045e65199c774a42027c53642210fe2e77fb56c5/packages/in_app_purchase/in_app_purchase_storekit/lib/src/types/app_store_purchase_details.dart#L77

Seems to be setting `pendingCompletePurchase` to false **only** if the status is _pending_.
Shouldn't it be the other way around?

I'll try to test this by relying on the `status` field instead, will report back.

**Comment by lukeirvin** [comment]
Any updates on this one?

I'm stuck in a similar situation on iOS where I am getting 99+ items in my Purchase Details List and they are all `restore`.

I've tried using `finishTransaction` before `_listenToPurchaseUpdated` but no luck.

So now, my device has a canceled subscription, but because my list is never empty I am stuck in a restore loop.

Is it possible there is just that much delay from Apple after calling `finishTransaction`? Or is there a better place to call this?

**Comment by mdmota** [comment]
> Any updates on this one?
> 
> I'm stuck in a similar situation on iOS where I am getting 99+ items in my Purchase Details List and they are all `restore`.
> 
> I've tried using `finishTransaction` before `_listenToPurchaseUpdated` but no luck.
> 
> So now, my device has a canceled subscription, but because my list is never empty I am stuck in a restore loop.
> 
> Is it possible there is just that much delay from Apple after calling `finishTransaction`? Or is there a better place to call this?

Do you really want a tip?
  use revenueCat, much easier to implement. worth the migration

**Comment by lukeirvin** [comment]
> > Any updates on this one?
> > I'm stuck in a similar situation on iOS where I am getting 99+ items in my Purchase Details List and they are all `restore`.
> > I've tried using `finishTransaction` before `_listenToPurchaseUpdated` but no luck.
> > So now, my device has a canceled subscription, but because my list is never empty I am stuck in a restore loop.
> > Is it possible there is just that much delay from Apple after calling `finishTransaction`? Or is there a better place to call this?
> 
> Do you really want a tip? use revenueCat, much easier to implement. worth the migration

I may end up doing so...

This seems to be the last pain point I'm trying to solve with `in_app_purchase` and I just can't figure it out.

**Comment by SamerOrfali22** [comment]
any update?

**Comment by ebsangam** [comment]
Anyone found any workaround or solution for this?

**Comment by deekshithdv** [comment]
I am also facing the same issue. When I tried to restore the purchases, it gives out all the purchases done and are over till now.
I don't find any purchases under subscriptions in the APP store though. But android works fine. Any idea on why IOS is behaving so?

**Comment by deekshithdv** [comment]
I kinda figured out the solution to this issue.

In IOS whenever you do iap.restorePurchases(), it will get all the past purchases irrespective of whether it is active or completed. Listener will catch these purchaseList which will be in the status of **restored**. 

Say for example if you done 50 purchases before, restorePurchases will get all 50 purchases as restored, so when it comes to listener you can't verify all 50 purchases with your APP's api. So it is better to fetch latest purchase and verify with the API. Then mark everything as purchaseCompleted else whenever you try to purchase a new plan or anything it will throw the following exception.

> [VERBOSE-2:ui_dart_state.cc(199)] Unhandled Exception: PlatformException(storekit_duplicate_product_object, There is a pending transaction for the same product identifier. Please either wait for it to be finished or finish it manually using `completePurchase` to avoid edge cases., {applicationUsername: null, requestData: null, quantity: 1, productIdentifier: one_month_subscription, simulatesAskToBuyInSandbox: false}, null)


Here is my code

``` 
_listenToPurchaseUpdated(
    List<PurchaseDetails> purchaseDetailsList,
  ) {
      for (final purchaseDetails in purchaseDetailsList.reversed) {
        if (purchaseDetails.status == PurchaseStatus.pending) {
           // show_PendingUI
        } 
        else {
            if (purchaseDetails.status == PurchaseStatus.error) {
              // Handle error
            } 
           else if (purchaseDetails.status == PurchaseStatus.purchased) {
              _verifyPurchaseDetails(purchaseDetails);
            } 
           else if (purchaseDetails.status == PurchaseStatus.restored) {
             // verify only the latest purchase
             if (purchaseDetails == purchaseDetailsList.last) {
               _verifyPurchaseDetails(purchaseDetails);
             }
           } 
          if (purchaseDetails.pendingCompletePurchase) {
            InAppPurchase.instance.completePurchase(purchaseDetails);
          }
       }
    }
  }


buyProduct() {
    final ProductDetails productDetails = ... // Saved earlier from queryProductDetails().
    final PurchaseParam purchaseParam = PurchaseParam(productDetails: productDetails);
    try {
          if (Platform.isIOS) {
           // complete the transactions if any...
            final transactions = await SKPaymentQueueWrapper().transactions();
            for (final skPaymentTransactionWrapper in transactions) {
                await SKPaymentQueueWrapper().finishTransaction(skPaymentTransactionWrapper);
            }
         }
        InAppPurchase.instance.buyNonConsumable(purchaseParam: purchaseParam)
        } catch (e) {
          // Handle error
        }
    }
}
```

Hope this would help. Cheers!!

**Comment by bfwg** [comment]
I found the same issue will happen for `PurchaseStatus.purchased` when you call `InAppPurchase.instance.buyNonConsumable` the second time to trigger the "You're currently subscribe to this." dialog. 
This is my workaround.
```dart
    purchaseDetailsList.forEach((PurchaseDetails purchaseDetails) async {
      if (purchaseDetails.status == PurchaseStatus.pending) {
        showPendingUI();
      } else {
        if (purchaseDetails.status == PurchaseStatus.error) {
          _handleError(purchaseDetails.error!);
        } else if ((purchaseDetails.status == PurchaseStatus.purchased ||
                purchaseDetails.status == PurchaseStatus.restored) &&
            purchaseDetailsList.length == 1) {
          bool valid = await _verifyPurchase(purchaseDetails);
          if (valid) {
            await _deliverProduct(purchaseDetails);
          } else {
            _handleInvalidPurchase(purchaseDetails);
          }
        } else if (purchaseDetails.status == PurchaseStatus.canceled) {
          EasyLoading.dismiss();
        }

        if (purchaseDetails.pendingCompletePurchase) {
          await InAppPurchase.instance.completePurchase(purchaseDetails);
        }
      }

```

**Comment by junixapp** [comment]
same issue ,

**Comment by adar2378** [comment]
Same issue, when I trigger restorePurchases methd. It returns tons of events via the purchaseStream and our server is crashing. Don't know what to do.

**Comment by hortigado** [comment]
> _subscription?.cancel();

Thanks, after hours of searching for the error, it was that of doing a cancel before starting another listen. In my case I had the cancel but without the await.

**Comment by Chay2203** [comment]
hey, is this resovled?

**Comment by fischja** [comment]
I have the same issue

**Comment by nathan1658** [comment]
same here

**Comment by thenishchalraj** [comment]
restored purchases are not clearing, been facing this issue for last 3 months. no resolution.
